### PR TITLE
ARROW-10406: [C++] Unify dictionaries when writing IPC file in a single shot

### PR DESCRIPTION
**Pull request by pitrou**
This is governed by a new option in IpcWriteOptions.
Unification is supported only for certain value types.
For unsupported types, it will always fail even when the dictionaries are equal.

### CODE REVIEW

**Review by bkietz — CHANGES_REQUESTED** [review]
Thanks for working on this!

Would it be worthwhile to expose table unification to python/R? Maybe as a trivial compute function?

File: cpp/src/arrow/ipc/writer.cc [hunk 1]
```diff
@@ -1001,6 +1001,17 @@ class ARROW_EXPORT IpcFormatWriter : public RecordBatchWriter {
...
 
+  Status WriteTable(const Table& table, int64_t max_chunksize) override {
+    std::shared_ptr<Table> owned_table;
+    const Table* table_ptr = &table;
+    if (is_file_format_ && options_.unify_dictionaries) {
+      ARROW_ASSIGN_OR_RAISE(owned_table,
+                            DictionaryUnifier::UnifyTable(table, options_.memory_pool));
+      table_ptr = owned_table.get();
+    }
+    return RecordBatchWriter::WriteTable(*table_ptr, max_chunksize);
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/ipc/writer.cc:1012]: Thanks for the suggestion, it's better indeed.

**Comment by pitrou** [comment]
> Would it be worthwhile to expose table unification to python/R? 

I can do it in Python.

File: cpp/src/arrow/array/concatenate.cc [hunk 2]
```diff
@@ -330,9 +318,8 @@ class ConcatenateImpl {
       return ConcatenateBuffers(index_buffers, pool_).Value(&out_->buffers[1]);
     } else {
       ARROW_ASSIGN_OR_RAISE(auto index_lookup, UnifyDictionaries(d));
-      DictionaryConcatenate concatenate(index_buffers, index_lookup, pool_);
-      RETURN_NOT_OK(VisitTypeInline(*d.index_type(), &concatenate));
-      out_->buffers[1] = std::move(concatenate.out_);
+      ARROW_ASSIGN_OR_RAISE(out_->buffers[1],
+                            ConcatenateDictionaryIndices(*fixed, index_lookup, pool_));
```
Comment (bkietz) [inline cpp/src/arrow/array/concatenate.cc:322]:
```suggestion
                            ConcatenateDictionaryIndices(*fixed, index_lookup));
```
No need to pass a member here

**Review by bkietz — APPROVED** [review]
LGTM

**Comment by nealrichardson** [comment]
I'm not sure this fully solves the issue. `ipc::feather::WriteTable` (what `write_feather()` calls) doesn't take `IpcWriteOptions`: https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/feather.cc#L792-L807

**Comment by pitrou** [comment]
Hmm, I'm not sure how to evolve the Feather APIs. They take their own options ("properties"). @wesm What do you think?

**Comment by wesm** [comment]
Can we just have `unify_dictionaries` always enabled when writing Feather files?

**Comment by pitrou** [comment]
> Can we just have `unify_dictionaries` always enabled when writing Feather files?

That sounds reasonable to me.

**Comment by pitrou** [comment]
While we're at it, should we also enable `allow_64bit` in Feather? AFAIK, Feather is only supported in C++ and associated languages (Python, R...).

**Comment by pitrou** [comment]
@bkietz Do you want to give this a second look since the Feather changes?

**Review by bkietz — APPROVED** [review]
LGTM, will merge

**Pull request closed without merging**